planner/core/task.go

```diff
@@ -682,6 +683,49 @@ func (p *basePhysicalAgg) newPartialAggregate() (partial, final PhysicalPlan) {
 		groupByItems = append(groupByItems, gbyCol)
 	}
 
+	// Remove unnecessary schema column.
+	// When the select column is same with the group by key, the column can be removed and gets value from the group by key.
+	// e.g
+	// select a, count(b) from t group by a;
+	// The schema is [firstrow(a), count(b), a]. The column firstrow(a) is unnecessary.
+	// Can optimize the schema to [count(b), a] , and change the index to get value.
+	partialCursor = 0
+	partialAggFuncs := make([]*aggregation.AggFuncDesc, 0, len(p.AggFuncs))
+	partialOptimizeSchema := expression.NewSchema()
+	for i, aggFunc := range p.AggFuncs {
+		if aggFunc.Name == ast.AggFuncFirstRow {
+			canOptimize := false
+			for j, gbyExpr := range p.GroupByItems {
+				if gbyExpr.Equal(p.ctx, aggFunc.Args[0]) {
+					canOptimize = true
+					finalAggFuncs[i].Args[0] = groupByItems[j]
+					break
+				}
+			}
+			if canOptimize {
+				partialCursor++
+				continue
+			}
+		}
+		if aggregation.NeedCount(aggFunc.Name) {
+			partialOptimizeSchema.Append(partialSchema.Columns[partialCursor])
+			partialCursor++
+		}
+		if aggregation.NeedValue(aggFunc.Name) {
+			partialOptimizeSchema.Append(partialSchema.Columns[partialCursor])
+			partialCursor++
+		}
+		partialAggFuncs = append(partialAggFuncs, aggFunc)
+	}
+	for i := range p.GroupByItems {
+		partialOptimizeSchema.Append(partialSchema.Columns[partialCursor+i])
+	}
+	p.AggFuncs = partialAggFuncs
+	p.schema = partialOptimizeSchema
+	for i, col := range p.schema.Columns {
+		col.ColName = model.NewCIStr(fmt.Sprintf("col_%d", i))
```

Comment:
I create all columns first, and then use the useful columns to build the new schema. Some columns are not used.
If i don't rename the column, the `explain` result maybe make some misleading(I think).
(Don't rename the column also can gets the right query result, the only difference is `explain` result)
e.g. 
```
mysql> desc select count(b),a,sum(b) from t group by a;
+-------------------------+----------+------+-----------------------------------------------------------------+
| id                      | count    | task | operator info                                                   |
+-------------------------+----------+------+-----------------------------------------------------------------+
| Projection_4            | 8000.00  | root | 2_col_0, test.t.a, 2_col_1                                      |
| └─HashAgg_9             | 8000.00  | root | group by:col_2, funcs:count(col_0), sum(col_1), firstrow(col_2) |
|   └─TableReader_10      | 8000.00  | root | data:HashAgg_5                                                  |
|     └─HashAgg_5         | 8000.00  | cop  | group by:test.t.a, funcs:count(test.t.b), sum(test.t.b)         |
|       └─TableScan_8     | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo      |
+-------------------------+----------+------+-----------------------------------------------------------------+
5 rows in set (0.01 sec)
```

If i don't rename the columns.
```
mysql> desc select count(b),a,sum(b) from t group by a;
+-------------------------+----------+------+-----------------------------------------------------------------+
| id                      | count    | task | operator info                                                   |
+-------------------------+----------+------+-----------------------------------------------------------------+
| Projection_4            | 8000.00  | root | 2_col_0, test.t.a, 2_col_1                                      |
| └─HashAgg_9             | 8000.00  | root | group by:col_3, funcs:count(col_0), sum(col_1), firstrow(col_3) |
|   └─TableReader_10      | 8000.00  | root | data:HashAgg_5                                                  |
|     └─HashAgg_5         | 8000.00  | cop  | group by:test.t.a, funcs:count(test.t.b), sum(test.t.b)         |
|       └─TableScan_8     | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo      |
+-------------------------+----------+------+-----------------------------------------------------------------+
5 rows in set (0.00 sec)
```
In `HashAgg_9` ,it group by `col_3`. But in `HashAgg_5` ,there are only 3 columns(`col_0`-`col2`).

Reply:
The name of those intermediate columns actually has no meaning, so IMHO the renaming seems unnecessary.

Reply:
**Just a suggestion**. How about using the same column name in the partial and final aggregate operators, i.e. `HashAgg_5` and `HashAgg_9`? So that users can notice the blood relationship between columns from different operators.